tests/es5.js

```diff
@@ -0,0 +1,202 @@
+    var blockHistory = ImmutableList([new MockBlock(0x7777)]);
+    var newBlock = new MockBlock(0x7779);
+    var blockRetention = 5;
+
+    reconcileBlockHistoryEs5(getBlockByHash, blockHistory, newBlock, onBlockAdded, onBlockRemoved, blockRetention, function (error, newHistory) {
+      expect(error).to.be.an("error");
+      expect(error.message).to.equal("apple");
+      done();
+    });
+  });
+
+  it("correctly rejects when getBlockByHash throws", function (done) {
+    var getBlockByHash = function (hash, callback) {
+      throw new Error("apple");
+    }
+    var blockHistory = ImmutableList([new MockBlock(0x7777)]);
+    var newBlock = new MockBlock(0x7779);
+    var blockRetention = 5;
+
+    reconcileBlockHistoryEs5(getBlockByHash, blockHistory, newBlock, onBlockAdded, onBlockRemoved, blockRetention, function (error, newHistory) {
+      expect(error).to.be.an("error");
+      expect(error.message).to.equal("apple");
+      done();
+    });
+  });
+
+  it("correctly rejects when onBlockAdded calls back with error", function (done) {
+    var getBlockByHash = function (hash, callback) { throw new Error("unreachable"); }
+    var failingOnBlockAdded = function (block, callback) { callback(new Error("apple")); }
+    var blockHistory = null;
+    var newBlock = new MockBlock(0x7777);
+    var blockRetention = 5;
+
+    reconcileBlockHistoryEs5(getBlockByHash, blockHistory, newBlock, failingOnBlockAdded, onBlockRemoved, blockRetention, function (error, newHistory) {
+      expect(error).to.be.an("error");
+      expect(error.message).to.equal("apple");
+      done();
+    });
+  });
+
+  it("accepts onBlockRemoved function that uses a callback", function (done) {
+    var getBlockByHash = getBlockByHashFactory();
+    var getBlockByHash = function (hash, callback) { throw new Error("unreachable"); }
+    var blockHistory = ImmutableList([
+      new MockBlock(0x7777, "AAAA", "AAAA"),
+      new MockBlock(0x7778, "AAAA", "AAAA"),
+    ]);
+    var newBlock = new MockBlock(0x7778, "BBBB", "AAAA");
+    var blockRetention = 5;
+
+    reconcileBlockHistoryEs5(getBlockByHash, blockHistory, newBlock, onBlockAdded, onBlockRemoved, blockRetention, function (error, newHistory) {
+      expect(error).to.be.undefined;
+      expect(newHistory.toJS()).to.deep.equal([
+        new MockBlock(0x7777, "AAAA", "AAAA"),
+        new MockBlock(0x7778, "BBBB", "AAAA"),
+      ]);
+      expect(newBlockAnnouncements).to.deep.equal([new MockBlock(0x7778, "BBBB", "AAAA")]);
+      expect(blockRemovalAnnouncments).to.deep.equal([new MockBlock(0x7778, "AAAA", "AAAA")]);
+      done();
+    });
+  });
+});
+
+describe("reconcileBlocksAndLogsEs5", function () {
+  var logAddedAnnonucements;
+  var logRemovedAnnouncements;
+  function onLogAdded(block, callback) { setTimeout(function () { logAddedAnnonucements.push(block); callback(); }); }
+  function onLogRemoved(block, callback) { setTimeout(function () { logRemovedAnnouncements.push(block); callback(); }); }
```

Comment:
The reconciler won't continue until these functions have finished doing their work.  In synchronous terms, you can think of it like these callbacks are made as blocking calls.

If the callback is called with an error result (or it throws), then that signals to the system that something went wrong with processing the stream and that it should consider this a failure and rollback.

This rollback has some pretty hefty caveats though, so the caller should avoid thorwing or calling the callback with an error unless things are really bad.

This pattern came out originally with the block-stream side of things where the thing processing the block needed a way to signal to the block stream that the block it just announced is somehow bad (e.g., the user can't fetch logs for the block or fetching the logs returns an incorrect result).  I kept the pattern for the log stream because it seemed reasonable that the user could similarly want to alert the log stream that something is horribly wrong and it should rollback.